EasyLM/models/llama/llama_train_ppo.py

```diff
@@ -0,0 +1,421 @@
+        value_train_state = value_train_state.apply_gradients(grads=value_grads)
+
+    return policy_train_state, value_train_state, loss
+
+
+def main(argv):
+    JaxDistributedConfig.initialize(FLAGS.jax_distributed)
+
+    variant = mlxu.get_user_flags(FLAGS, FLAGS_DEF)
+    flags_config_dict = mlxu.user_flags_to_config_dict(FLAGS, FLAGS_DEF)
+    logger = mlxu.WandBLogger(
+        config=FLAGS.logger,
+        variant=variant,
+        enable=FLAGS.log_all_worker or (jax.process_index() == 0),
+    )
+    set_random_seed(FLAGS.seed)
+
+    tokenizer = LLaMAConfig.get_tokenizer(FLAGS.tokenizer)
+    dataset = DatasetFactory.load_dataset(FLAGS.train_dataset, tokenizer)
+    if FLAGS.load_dataset_state != '':
+        dataset.load_state_dict(mlxu.load_pickle(FLAGS.load_dataset_state))
+
+    if isinstance(dataset, torch.utils.data.DataLoader):
+        wrapped_dataset = dataset.dataset
+    else:
+        wrapped_dataset = dataset
+
+    real_batch_size = wrapped_dataset.config.batch_size
+    # for the scheduler, which only gets updated with 'real' grad steps
+    simulated_batch_size = real_batch_size * FLAGS.optimizer.accumulate_gradient_steps
+    steps_per_epoch = len(wrapped_dataset) // real_batch_size
+    simulated_steps_per_epoch = len(wrapped_dataset) // simulated_batch_size
+    print(f"Make sure your scheduler steps are based on the simulated batch size: {simulated_batch_size}!")
+    print(f"Total simulated steps: {simulated_steps_per_epoch * FLAGS.num_epochs}")
+
+    seq_length = wrapped_dataset.seq_length
+
+    print("Building model...")
+    if FLAGS.load_llama_config != '':
+        llama_config = LLaMAConfig.load_config(FLAGS.load_llama_config)
+    else:
+        llama_config = LLaMAConfig(**FLAGS.llama)
+
+    if FLAGS.update_llama_config != '':
+        llama_config.update(dict(eval(FLAGS.update_llama_config)))
+
+    llama_config.update(dict(
+        bos_token_id=wrapped_dataset.tokenizer.bos_token_id,
+        eos_token_id=wrapped_dataset.tokenizer.eos_token_id,
+    ))
+    if llama_config.vocab_size < wrapped_dataset.vocab_size:
+        llama_config.update(dict(vocab_size=wrapped_dataset.vocab_size))
+
+    _policy_model = FlaxLLaMAForCausalLMModule(
+        llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype)
+    )
+    # TODO: load from pretrained ckpt
+    # policy_model = FlaxLLaMAForCausalLM.from_pretrained('meta-llama/Llama-2-7b-hf', config=llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype), from_pt=True, use_auth_token=True)
+    policy_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
+    value_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
+    reference_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
+    reward_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
```

Comment:
It might be possible to re-use this CausalLM class for RM, where we take the 0-th logit as the reward. After all, lm_head is a DxV linear layer, while the reward_head is a Dx1 linear layer. This shouldn't be too much overhead ...

Reply:
Yeah, agree there. I think that having a reward class that wraps this behaviour is still a good idea, though, since then it means adding other reward model implementations (e.g., you generate a score or ping gpt-4 or something weird) is easier.